src/pudl_rmi/formatter_optimus.py

```diff
@@ -0,0 +1,251 @@
+    "faked_66": "CWIP_2042",
+    "faked_67": "CWIP_2043",
+    "faked_68": "CWIP_2044",
+    "faked_69": "CWIP_2045",
+    "faked_70": "CWIP_2046",
+    "faked_71": "CWIP_2047",
+    "faked_72": "CWIP_2048",
+    "faked_73": "CWIP_2049",
+    "faked_74": "CWIP_2050",
+    "line_id": "Record ID Depreciation",
+    "record_id_ferc1": "Record ID FERC 1",
+    "plant_id_eia": "Plant ID EIA",
+    "report_year": "Report Year",
+    "data_source": "Data Source of Depreciation Study",
+    "ferc_acct_name": "FERC Acct",
+}
+
+TECHNOLOGY_DESCRIPTION_TO_RESOURCE_TYPE = {
+    "Conventional Steam Coal": "Coal",
+    "Natural Gas Fired Combined Cycle": "NaturalGasCC",
+    "Natural Gas Fired Combustion Turbine": "NaturalGasCT",
+    "Natural Gas Steam Turbine": "NaturalGasCT",  # MAYBE?!?
+    "Geothermal": "Geothermal",
+    "Onshore Wind Turbine": "LandbasedWind",
+    "Conventional Hydroelectric": "Hydropower",
+    # pd.NA: 'Transmission',
+    # pd.NA: 'Distribution',
+    "Solar Photovoltaic": "UtilityPV",
+    "Nuclear": "Nuclear",
+    "Offshore Wind Turbine": "OffshoreWind",  # THIS ONE IS A GUESS
+    "Solar Thermal with Energy Storage": "SolarPlusBattery",
+    # pd.NA: 'EE',
+    # pd.NA: 'DR',
+    # pd.NA: 'Battery'
+}
+
+UTILITY_RENAME: Dict = {
+    "Carolina Power & Light Co": "Duke Energy Progress",
+    "Duke Energy Corp": "Duke Energy Carolinas",
+}
+"""
+EIA utility names (keys) to Optimus-preferred names (values).
+"""
+
+
+def execute(
+    deprish_ferc1_eia: pd.DataFrame,
+    plants_eia860: pd.DataFrame,
+    utils_eia860: pd.DataFrame,
+    **kwargs,
+) -> pd.DataFrame:
+    """
+    Convert RMI outputs into optimus model inputs.
+
+    This function converts the Deprish-FERC-EIA output to the RMI Optimus format.
+    This conversion mostly consists of renaming columns and content. It also
+    adds two data columns from PUDL tables: ``state`` from the ``plants_entity_eia``
+    table and ``utility_id_pudl`` from ``utilities_eia``.
+
+    Args:
+        deprish_ferc1_eia: table of depreciation & FERC 1 & EIA data.
+        plants_eia860: ``pudl_out.plants_eia860()``
+        utils_eia860: ``pudl_out.utils_eia860()``
+        kwargs: Additional kwargs to be passed into :func:``select_from_deprish_ferc1``
+            if you want to select specific utilities/years/data sources.
+    """
+    # if there are kwargs, this means you want to select from deprish_ferc1_eia
+    if kwargs:
+        deprish_ferc1_eia = select_from_deprish_ferc1(deprish_ferc1_eia, **kwargs)
+    # add plant state and the EIA utility name
+    model_input = (
+        deprish_ferc1_eia.reset_index()  # reset index b4 merge bc it's prob 'record_id_eia'
+        .merge(
+            plants_eia860[["plant_id_eia", "report_date", "state"]].drop_duplicates(),
+            on=["plant_id_eia", "report_date"],
+            validate="m:1",
+            how="left",
+        )
+        .merge(
+            utils_eia860.sort_values(["report_date"], ascending=False)[
```

Comment:
Does the merge care if you sort the values in the right dataframe? If it is a left merge I would assume it just keeps the rows in whatever order the left df is.

Reply:
Ah yes i should leave a comment in here about why I'm doing that.... bc it took me a minute to remember why. I'm sorting by report date to get the most recent utility name because they can change over time and rmi prefers the up-to-date names.